### PR TITLE
Integrate IREE C/C++ source at c4624870 into samples.

### PR DESCRIPTION
**Pull request by ScottTodd**
https://github.com/iree-org/iree/commit/c4624870a8de85224c3f02f4ea8e26837cd99e2f

* Bump C++ standard from 14 to 17
* Use "llvm-cpu" target naming
* Use new hal/drivers/* paths for Vulkan driver
* Adapt to HAL and VM API changes

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -28,9 +28,9 @@ include(FetchContent)
 FetchContent_Declare(
   iree
   GIT_REPOSITORY https://github.com/iree-org/iree.git
-  GIT_TAG 7ca0f46fc38c7d5ccb7ee275db59ed7d992436ef # 2022-06-09
+  GIT_TAG c4624870a8de85224c3f02f4ea8e26837cd99e2f # 2022-08-12
   GIT_SUBMODULES_RECURSE OFF
-  GIT_SHALLOW ON
+  GIT_SHALLOW OFF
```
Comment (jpienaar) [inline cpp/CMakeLists.txt:33]: Why this change?

Comment (ScottTodd) [reply cpp/CMakeLists.txt:33]: CMake wasn't actually fetching the specified commit without it. It might work for the tip of a branch, but I have this pinned to an older commit without a specific branch.

**Review by benvanik — APPROVED** [review]
